File: packages/core-ts/src/paths.ts
/**
 * Source authorization: roots, canonicalization, file identity and secret policy.
 *
 * Nothing becomes a source unless it canonicalizes inside a configured workspace root, is
 * a regular file, and survives the secret policy. Rejections carry a code and a bounded
 * token - never the path, and never the matched value.
 */
import { lstatSync, openSync, readSync, closeSync, realpathSync, statSync } from "node:fs";
import { basename, extname, isAbsolute, resolve, sep } from "node:path";

import { ShuntError } from "./errors.js";

const SECRET_NAMES = new Set([
  ".env", ".env.local", ".env.production", ".env.development", ".netrc", "_netrc",
  "credentials", "auth.json", ".htpasswd", "id_rsa", "id_dsa", "id_ecdsa", "id_ed25519",
  ".pgpass", "shadow", "master.key", "secrets.yaml", "secrets.yml", ".npmrc", ".pypirc",
  ".dockercfg",
]);
const SECRET_SUFFIXES = ["_rsa", "_dsa", "_ed25519"];
const SECRET_EXTENSIONS = new Set([".pem", ".key", ".p12", ".pfx", ".jks", ".keystore", ".ppk"]);
const SECRET_DIR_PARTS = new Set([".ssh", ".gnupg", ".aws", ".kube", ".docker"]);

export interface PathPolicy {
  readonly roots: readonly string[];
  readonly denylist: readonly string[];
}

export function pathPolicy(roots: readonly string[], denylist: readonly string[] = []): PathPolicy {
  if (roots.length === 0) throw new ShuntError("UNSAFE_SOURCE", "NO_WORKSPACE_ROOT");
  // Roots are canonicalized the same way sources are, so a root reached through a
  // symlinked prefix (macOS /var -> /private/var, for one) still contains its own files.
  return { roots: roots.map(canonicalRoot), denylist };
}

function canonicalRoot(root: string): string {
  try {
    return realpathSync(resolve(root));
  } catch {
    return resolve(root);
  }
}

export interface AuthorizedPath {
  readonly real: string;
  readonly dev: number;
  readonly ino: number;
  readonly size: number;
  readonly mtimeNs: bigint;
}

function isSecretPath(real: string, policy: PathPolicy): boolean {
  const name = basename(real).toLowerCase();
  if (SECRET_NAMES.has(name)) return true;
  if (SECRET_EXTENSIONS.has(extname(real).toLowerCase())) return true;
  if (SECRET_SUFFIXES.some((sfx) => name.endsWith(sfx))) return true;
  if (name.startsWith(".env")) return true;
  return real.split(sep).some((part) => SECRET_DIR_PARTS.has(part));
}

function matchesDenylist(real: string, policy: PathPolicy): boolean {
  for (const root of policy.roots) {
    if (!real.startsWith(root + sep)) continue;
    const rel = real.slice(root.length + 1);
    for (const pattern of policy.denylist) {
      const rx = new RegExp(
        "^" + pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, "[^/]*") + "$",
      );
      if (rx.test(rel)) return true;
      const dir = rel.split("/")[0];
      if (dir !== undefined && rx.test(`${dir}/*`)) return true;
    }
  }
  return false;
}

/** Canonicalize and authorize one source path, or throw a bounded ShuntError. */
export function authorize(path: string, policy: PathPolicy): AuthorizedPath {
  if (!isAbsolute(path)) throw new ShuntError("UNSAFE_SOURCE", "RELATIVE_PATH");
  let lst;
  try {
    lst = lstatSync(path, { bigint: true });
  } catch {
    throw new ShuntError("UNSAFE_SOURCE", "NOT_FOUND");
  }
  if (lst.isSymbolicLink()) throw new ShuntError("UNSAFE_SOURCE", "SYMLINK");

  let real: string;
  try {
    real = realpathSync(path);
  } catch {
    real = resolve(path);
  }
  const inRoot = policy.roots.some((root) => real === root || real.startsWith(root + sep));
  if (!inRoot) throw new ShuntError("UNSAFE_SOURCE", "OUTSIDE_WORKSPACE_ROOT");
  if (isSecretPath(real, policy) || matchesDenylist(real, policy)) {
    throw new ShuntError("UNSAFE_SOURCE", "SECRET_PATH");
  }

  let st;
  try {
    st = statSync(real, { bigint: true });
  } catch {
    throw new ShuntError("UNSAFE_SOURCE", "NOT_FOUND");
  }
  if (!st.isFile()) throw new ShuntError("UNSAFE_SOURCE", "NOT_REGULAR_FILE");
  if (st.nlink > 1n) {
    // A hardlinked file can be re-pointed outside the root between checks.
    throw new ShuntError("UNSAFE_SOURCE", "HARDLINKED");
  }
  return {
    real,
    dev: Number(st.dev),
    ino: Number(st.ino),
    size: Number(st.size),
    mtimeNs: st.mtimeNs,
  };
}

/** Stream a file, refusing at the cap instead of allocating past it. */
export function readBounded(path: string, maxBytes: number): Uint8Array {
  const CHUNK = 256 * 1024;
  const fd = openSync(path, "r");
  const parts: Uint8Array[] = [];
  let total = 0;
  try {
    for (;;) {
      const buf = Buffer.allocUnsafe(CHUNK);
      const read = readSync(fd, buf, 0, CHUNK, null);
      if (read === 0) break;
      if (total + read > maxBytes) throw new ShuntError("LIMIT_EXCEEDED", "SOURCE_OVER_BYTE_CAP");
      parts.push(buf.subarray(0, read));
      total += read;
    }
  } finally {
    closeSync(fd);
  }
  return Buffer.concat(parts, total);
}

/**
 * Snapshot a file and re-check identity afterwards: a file swapped mid-read yields
 * `SOURCE_CHANGED` rather than a snapshot mixing two versions.
 */
export function assertUnchanged(authorized: AuthorizedPath, readBytes: number): void {
  let after;
  try {
    after = statSync(authorized.real, { bigint: true });
  } catch {
    throw new ShuntError("SOURCE_CHANGED", "DISAPPEARED");
  }
  if (Number(after.dev) !== authorized.dev || Number(after.ino) !== authorized.ino) {
    throw new ShuntError("SOURCE_CHANGED", "IDENTITY_CHANGED");
  }
  if (Number(after.size) !== readBytes || after.mtimeNs !== authorized.mtimeNs) {
    throw new ShuntError("SOURCE_CHANGED", "MODIFIED_DURING_READ");
  }
}

File: packages/core-py/src/context_shunt/paths.py
"""Source authorization: roots, canonicalization, file identity and secret policy.

Nothing becomes a source unless it canonicalizes inside a configured workspace root,
is a regular file, and survives the secret policy. Rejections carry a code and a
bounded token - never the path, and never the matched value.
"""

from __future__ import annotations

import os
import stat
from dataclasses import dataclass, field
from pathlib import Path, PurePosixPath

from .errors import ShuntError

# Filenames and suffixes that are refused outright. Detection is best effort: this is a
# denylist, not a proof of secrecy, which is why the provider must also be an approved
# data destination and why chunks stay minimal.
_SECRET_NAMES = frozenset(
    {
        ".env",
        ".env.local",
        ".env.production",
        ".env.development",
        ".netrc",
        "_netrc",
        "credentials",
        "auth.json",
        ".htpasswd",
        "id_rsa",
        "id_dsa",
        "id_ecdsa",
        "id_ed25519",
        ".pgpass",
        "shadow",
        "master.key",
        "secrets.yaml",
        "secrets.yml",
        ".npmrc",
        ".pypirc",
        ".dockercfg",
    }
)
_SECRET_SUFFIXES = ("_rsa", "_dsa", "_ed25519")
_SECRET_EXTENSIONS = frozenset({".pem", ".key", ".p12", ".pfx", ".jks", ".keystore", ".ppk"})
_SECRET_DIR_PARTS = frozenset({".ssh", ".gnupg", ".aws", ".kube", ".docker"})

# Content markers checked on snapshot bytes, questions, answers and quotes.
_SECRET_CONTENT_MARKERS = (
    b"-----BEGIN RSA PRIVATE KEY-----",
    b"-----BEGIN OPENSSH PRIVATE KEY-----",
    b"-----BEGIN DSA PRIVATE KEY-----",
    b"-----BEGIN EC PRIVATE KEY-----",
    b"-----BEGIN PGP PRIVATE KEY BLOCK-----",
    b"-----BEGIN PRIVATE KEY-----",
    b"aws_secret_access_key",
    b"AKIA",
    b"ghp_",
    b"github_pat_",
    b"xoxb-",
    b"xoxp-",
    b"sk-ant-",
)


@dataclass(frozen=True)
class PathPolicy:
    """Workspace roots plus an administrator denylist of relative globs."""

    roots: tuple[Path, ...]
    denylist: tuple[str, ...] = ()
    follow_symlinks: bool = False
    extra_secret_names: frozenset[str] = field(default_factory=frozenset)

    @classmethod
    def from_config(cls, roots: list[str], denylist: list[str] | None = None) -> PathPolicy:
        if not roots:
            raise ShuntError("UNSAFE_SOURCE", "NO_WORKSPACE_ROOT")
        resolved = tuple(Path(r).resolve(strict=False) for r in roots)
        return cls(roots=resolved, denylist=tuple(denylist or ()))


@dataclass(frozen=True)
class AuthorizedPath:
    """A path proven safe, together with the identity used to detect a swap mid-read."""

    real: Path
    dev: int
    ino: int
    size: int
    mtime_ns: int

    def identity(self) -> tuple[int, int, int, int]:
        return (self.dev, self.ino, self.size, self.mtime_ns)


def _is_secret_path(real: Path, policy: PathPolicy) -> bool:
    name = real.name
    lowered = name.lower()
    if lowered in _SECRET_NAMES or lowered in policy.extra_secret_names:
        return True
    if real.suffix.lower() in _SECRET_EXTENSIONS:
        return True
    if any(lowered.endswith(sfx) for sfx in _SECRET_SUFFIXES):
        return True
    if lowered.startswith(".env"):
        return True
    return any(part in _SECRET_DIR_PARTS for part in real.parts)


def _matches_denylist(real: Path, policy: PathPolicy) -> bool:
    for root in policy.roots:
        try:
            rel = real.relative_to(root)
        except ValueError:
            continue
        posix = PurePosixPath(rel.as_posix())
        for pattern in policy.denylist:
            if posix.match(pattern):
                return True
    return False


def authorize(path: str, policy: PathPolicy) -> AuthorizedPath:
    """Canonicalize and authorize one source path, or raise a bounded ShuntError.

    ``os.lstat`` is used before resolution so a symlink is seen as a symlink, and the
    descriptor identity captured here is re-checked after reading to detect a swap.
    """
    candidate = Path(path)
    if not candidate.is_absolute():
        raise ShuntError("UNSAFE_SOURCE", "RELATIVE_PATH")
    try:
        lst = os.lstat(candidate)
    except OSError:
        raise ShuntError("UNSAFE_SOURCE", "NOT_FOUND") from None
    if stat.S_ISLNK(lst.st_mode) and not policy.follow_symlinks:
        raise ShuntError("UNSAFE_SOURCE", "SYMLINK")

    real = candidate.resolve(strict=False)
    if not any(real == root or root in real.parents for root in policy.roots):
        raise ShuntError("UNSAFE_SOURCE", "OUTSIDE_WORKSPACE_ROOT")
    if _is_secret_path(real, policy) or _matches_denylist(real, policy):
        raise ShuntError("UNSAFE_SOURCE", "SECRET_PATH")

    try:
        st = os.stat(real)
    except OSError:
        raise ShuntError("UNSAFE_SOURCE", "NOT_FOUND") from None
    if not stat.S_ISREG(st.st_mode):
        raise ShuntError("UNSAFE_SOURCE", "NOT_REGULAR_FILE")
    if st.st_nlink > 1:
        # A hardlinked file can be re-pointed outside the root between checks.
        raise ShuntError("UNSAFE_SOURCE", "HARDLINKED")
    return AuthorizedPath(
        real=real, dev=st.st_dev, ino=st.st_ino, size=st.st_size, mtime_ns=st.st_mtime_ns
    )


def contains_secret_marker(data: bytes) -> bool:
    return any(marker in data for marker in _SECRET_CONTENT_MARKERS)


def assert_no_secret(data: bytes, stage: str) -> None:
    """Reject the affected operation without echoing the matched value."""
    if contains_secret_marker(data):
        raise ShuntError("UNSAFE_SOURCE", f"SECRET_IN_{stage}")
